Source/drlg_l1.cpp

```diff
@@ -732,71 +641,68 @@ void ApplyShadowsPatterns()
 	}
 }
 
-bool PlaceMiniSet(const BYTE *miniset, int tmin, int tmax, int cx, int cy, bool setview)
+bool PlaceMiniSet(const BYTE *miniset, bool setview)
 {
-	int sx;
-	int sy;
-
 	int sw = miniset[0];
 	int sh = miniset[1];
+	int sx = GenerateRnd(DMAXX - sw) - 1;
+	int sy = GenerateRnd(DMAXY - sh);
 
-	int numt = 1;
-	if (tmax - tmin != 0) {
-		numt = GenerateRnd(tmax - tmin) + tmin;
-	}
-
-	for (int i = 0; i < numt; i++) {
-		sx = GenerateRnd(DMAXX - sw);
-		sy = GenerateRnd(DMAXY - sh);
-		bool abort = false;
-		int found = 0;
+	for (int bailcnt = 0;; bailcnt++) {
+		if (bailcnt > 4000)
+			return false;
 
-		while (!abort) {
-			abort = true;
-			if (cx != -1 && sx >= cx - sw && sx <= cx + 12) {
-				sx++;
-				abort = false;
-			}
-			if (cy != -1 && sy >= cy - sh && sy <= cy + 12) {
-				sy++;
-				abort = false;
+		sx++;
+		if (sx == DMAXX - sw) {
+			sx = 0;
+			sy++;
+			if (sy == DMAXY - sh) {
+				sy = 0;
 			}
+		}
 
-			int ii = 2;
+		if (SetPiecesRoom.Contains({ sx, sy })) {
+			continue;
+		}
 
-			for (int yy = 0; yy < sh && abort; yy++) {
-				for (int xx = 0; xx < sw && abort; xx++) {
-					if (miniset[ii] != 0 && dungeon[xx + sx][sy + yy] != miniset[ii])
-						abort = false;
-					if (Protected[xx + sx][sy + yy])
-						abort = false;
-					ii++;
-				}
-			}
+		// Limit the position of SetPieces for compatibility with Diablo bug
+		bool valid = true;
+		if (sx <= 12) {
+			sx++;
+			valid = false;
+		}
+		if (sy <= 12) {
+			sy++;
+			valid = false;
+		}
+		if (!valid) {
+			continue;
+		}
 
-			if (!abort) {
-				sx++;
-				if (sx == DMAXX - sw) {
-					sx = 0;
-					sy++;
-					if (sy == DMAXY - sh) {
-						sy = 0;
-					}
-				}
-				if (++found > 4000)
-					return false;
+		int ii = 2;
+
+		bool success = true;
+		for (int yy = 0; yy < sh && success; yy++) {
+			for (int xx = 0; xx < sw && success; xx++) {
+				if (miniset[ii] != 0 && dungeon[xx + sx][sy + yy] != miniset[ii])
+					success = false;
+				if (Protected[xx + sx][sy + yy])
+					success = false;
+				ii++;
 			}
 		}
+		if (success)
+			break;

```

Comment:
`Miniset::matches({sx, sy})` would be usable here

Reply:
That was the planned next step :)